### PR TITLE
NEW Use embed/embed v4

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/10243

Requires https://github.com/silverstripe/silverstripe-framework/pull/10244

### CODE REVIEW

File: code/Exceptions/InvalidUrlException.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace SilverStripe\AssetAdmin\Exceptions;
+
+use Exception;
+
+class InvalidUrlException extends Exception
```
Comment (maxime-rainville) [inline code/Exceptions/InvalidUrlException.php:7]: Why are we creating this exception?

File: code/Forms/RemoteFileFormFactory.php [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 namespace SilverStripe\AssetAdmin\Forms;
 
-use Embed\Exceptions\InvalidUrlException;
 use InvalidArgumentException;
+use SilverStripe\AssetAdmin\Exceptions\InvalidUrlException;
```
Comment (maxime-rainville) [inline code/Forms/RemoteFileFormFactory.php:6]: Seems like we should be swapping this for `NetworkException` or `RequestException`?

File: code/Forms/RemoteFileFormFactory.php [hunk 3]
```diff
@@ -2,8 +2,8 @@
```
Comment (maxime-rainville) [inline code/Forms/RemoteFileFormFactory.php:2]: Is this thing use when we hotlink to a file?

**Review by maxime-rainville — CHANGES_REQUESTED** [review]

File: code/Forms/RemoteFileFormFactory.php (hunk 3 above)
Comment (emteknetnz) [reply code/Forms/RemoteFileFormFactory.php:2]: Yes?

File: code/Forms/RemoteFileFormFactory.php (hunk 2 above)
Comment (emteknetnz) [reply code/Forms/RemoteFileFormFactory.php:6]: The RequestException implementation in embed/embed doesn't actually allow for custom messages, despite it being int he constructor.  We have a bunch of existing _t() data that we should retain.

https://github.com/oscarotero/Embed/blob/master/src/Http/RequestException.php#L16

File: code/Exceptions/InvalidUrlException.php (hunk 1 above)
Comment (emteknetnz) [reply code/Exceptions/InvalidUrlException.php:7]: Commented in reply below

File: code/Forms/RemoteFileFormFactory.php (hunk 2 above)
Comment (emteknetnz) [reply code/Forms/RemoteFileFormFactory.php:6]: Also the RequestException has 3 params, int $code and RequestInterface $request

We could make up a code, but we have no way to get $request from many contexts, eg

**RemoteFileFormFactory.php**

```php
    /**
     * Checks if the embed generated is not just a link
     *
     * @param Embeddable $embed
     * @return bool
     * @throws InvalidUrlException
     */
    protected function validateEmbed(Embeddable $embed)
    {
        if (!$embed->validate()) {
            throw new InvalidUrlException(_t(
                __CLASS__.'.ERROR_EMBED',
                'There is currently no embeddable media available from this URL'
            ));
        }
        return true;
    }
```

Comment (maxime-rainville) [reply code/Forms/RemoteFileFormFactory.php:6]: Cool ... that makes sense.

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
`InvalidUrlException` sound too generic for something that is very `oEmbed` specific. At the very least, there should be a comment on the class to explain its purpose.

I'm wondering if it wouldn't make more sense to have this exception in framework instead?

**Comment by emteknetnz** [comment]
Have renamed to `InvalidRemoteUrlException`

Note there's something of an unavoidable (semver breaking?) API change in that the Exception thrown has changed.  It's unavoidable because the old Exception came from embed/embed v3, which no longer exists.

File: code/Exceptions/InvalidRemoteUrlException.php [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace SilverStripe\AssetAdmin\Exceptions;
+
+use Exception;
+
+class InvalidRemoteUrlException extends Exception
```
Comment (maxime-rainville) [inline code/Exceptions/InvalidRemoteUrlException.php:7]: Can we include a comment to explains what this is for?

**Review by maxime-rainville — CHANGES_REQUESTED** [review]

File: code/Exceptions/InvalidRemoteUrlException.php (hunk 4 above)
Comment (emteknetnz) [reply code/Exceptions/InvalidRemoteUrlException.php:12]: Added

**Review by maxime-rainville — APPROVED** [review]

**Comment by maxime-rainville** [comment]
Build is failing because it's still trying to install an older version of frameworktest.

I took the liberty to bump it up

**Comment by maxime-rainville** [comment]
Actually, the framewortest version got fixed in https://github.com/silverstripe/silverstripe-asset-admin/pull/1259 already. I removed my extra commit, merged up the other fixed and rebased.

**Pull request merged**